capellambse/extensions/validation/_validate.py

```diff
@@ -0,0 +1,483 @@
+            category = Category[category]
+        return self[category]
+
+    def by_type(self, type: type[c.GenericElement] | str) -> Rules:
+        """Filter the validation results by ``type``."""
+        if not isinstance(type, str):
+            type = type.__name__
+        return Rules(
+            {
+                category: RuleList(
+                    rule for rule in rules if type in rule.types
+                )
+                for category, rules in self.items()
+                if rules
+            }
+        )
+
+
+VALIDATION_RULES = Rules(
+    {
+        Category.REQUIRED: RuleList([]),
+        Category.RECOMMENDED: RuleList([]),
+        Category.SUGGESTED: RuleList([]),
+    }
+)
+
+
+class Results(dict[Rule, dict[helpers.UUIDString, Result]]):
+    """A validation rule result mapping."""
+
+    def __getitem__(self, key: Rule | str) -> dict[helpers.UUIDString, Result]:
+        return super().__getitem__(key)  # type: ignore[index]
+
+    def by_uuid(self, uuid: helpers.UUIDString) -> dict[Rule, Result]:
+        """Filter the validation results by ``uuid``."""
+        return {rule: res[uuid] for rule, res in self.items() if uuid in res}
+
+    def by_category(self, category: Category | str) -> Results:
+        """Filter the validation results by ``category``."""
+        return Results(
+            {
+                rule: {
+                    uid: r
+                    for uid, r in res.items()
+                    if r.category == category and r
+                }
+                for rule, res in self.items()
+            }
+        )
+
+    def by_value(self, value: bool) -> Results:
+        """Filter the validation results by ``value``."""
+        return Results(
+            {
+                rule: {
+                    uid: r for uid, r in res.items() if r.value == value and r
+                }
+                for rule, res in self.items()
+            }
+        )
+
+    def by_type(self, type: str) -> Results:
+        """Filter the validation results by ``type``."""
+        results = dict[Rule, dict[helpers.UUIDString, Result]]()
+        for rule, res in self.items():
+            if not res:
+                continue
+
+            for rule_type in rule.types:
+                if rule_type != type:
+                    continue
+
+                results[rule] = res
+
+        return Results(results)
+
+    def setdefault(
+        self,
+        key: t.Any,
+        default: dict[helpers.UUIDString, Result] | None = None,
+    ) -> dict[helpers.UUIDString, Result]:
+        if not isinstance(default, dict):
+            raise ValueError("Only a result dictionary is accepted")
+        return super().setdefault(key, default)
+
+
+def register_rule(
+    category: Category,
+    types: (
+        str
+        | VirtualType[_T]
+        | type[_T]
+        | cabc.Iterable[str | VirtualType[_T] | type[_T]]
+    ),
+    id: str,
+    name: str,
+    rationale: str,
```

Comment:
The `rationale` can become quite long, and is currently always passed using implicitly concatenated string fragments. For ergonomic reasons, IMO it would make sense to allow passing a triple-quoted multi-line string (just like docstrings), which gets cleaned up with something like `inspect.cleandoc`.